### ISSUE TITLE
Celery job irods.connection issues in production

### ISSUE THREAD

**Issue by mikkonie** [bihealth/sodar-server#1535]
This seems to be the root cause for #1532 and #1533, plus some other recent issues.

During a few short timespans, Celery jobs have failed to complete with `irods.connection` problems. This has caused some issues with e.g. landing zone jobs failing without the appropriate revert actions taken. This has happened twice within the last month, in both cases for a <1h time span.

This appears to be a problem between python-irodsclient (used by celeryd) connecting to the iRODS iCAT server. The iCAT server and celeryd are inside the same local docker-compose network, so external network issues technically shouldn't cause connection problems, right?

iRODS logs do not seem to show anything related at this time, even failed connection attempts. There is also no error messages about e.g. too many open connections blocking iRODS access.

These issues have resolved themselves after a while with no need for e.g. container restarts, which would also speak in favor of some kind of a network problem.

Sentry says:

```
irods.connection    The read operation timed out
message             Could not receive server response
```

Celeryd error dump in comments below.

**Comment by mikkonie** [comment]
Celeryd dump:

```
sodar-celeryd-default_1   [2022-11-17 09:15:21,871: CRITICAL/ForkPoolWorker-15] The read operation timed out
sodar-celeryd-default_1   [2022-11-17 09:15:21,882: ERROR/ForkPoolWorker-15] Could not receive server response
sodar-celeryd-default_1   [2022-11-17 09:15:21,891: ERROR/ForkPoolWorker-15] Engine execution has failed, something bad must have happened (last 10 machine transitions were [('SCHEDULING', 'W
AITING'), ('WAITING', 'ANALYZING'), ('ANALYZING', 'SCHEDULING'), ('SCHEDULING', 'WAITING'), ('WAITING', 'ANALYZING'), ('ANALYZING', 'SCHEDULING'), ('SCHEDULING', 'WAITING'), ('WAITING', 'ANALYZING'), ('ANA
LYZING', 'GAME_OVER'), ('GAME_OVER', 'FAILURE')])
sodar-celeryd-default_1   Traceback (most recent call last):
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/message/__init__.py", line 171, in _recv_message_in_len
sodar-celeryd-default_1       buf = sock.recv(size_left, socket.MSG_WAITALL)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/ssl.py", line 1223, in recv
sodar-celeryd-default_1       raise ValueError(
sodar-celeryd-default_1   ValueError: non-zero flags not allowed in calls to recv() on <class 'ssl.SSLSocket'>
sodar-celeryd-default_1   
sodar-celeryd-default_1   During handling of the above exception, another exception occurred:
sodar-celeryd-default_1   
sodar-celeryd-default_1   Traceback (most recent call last):
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/connection.py", line 109, in recv
sodar-celeryd-default_1       msg = iRODSMessage.recv(self.socket)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/message/__init__.py", line 261, in recv
sodar-celeryd-default_1       rsp_header_size = _recv_message_in_len(sock, 4)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/message/__init__.py", line 173, in _recv_message_in_len
sodar-celeryd-default_1       buf = sock.recv(size_left)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/ssl.py", line 1226, in recv
sodar-celeryd-default_1       return self.read(buflen)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/ssl.py", line 1101, in read
sodar-celeryd-default_1       return self._sslobj.read(len)
sodar-celeryd-default_1   socket.timeout: The read operation timed out
sodar-celeryd-default_1   
sodar-celeryd-default_1   During handling of the above exception, another exception occurred:
sodar-celeryd-default_1   
sodar-celeryd-default_1   Traceback (most recent call last):
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/taskflow/engines/action_engine/engine.py", line 300, in run_iter
sodar-celeryd-default_1       failure.Failure.reraise_if_any(memory.failures)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/taskflow/types/failure.py", line 338, in reraise_if_any
sodar-celeryd-default_1       failures[0].reraise()
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/taskflow/types/failure.py", line 350, in reraise
sodar-celeryd-default_1       raise value
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/taskflow/engines/action_engine/executor.py", line 71, in _revert_task
sodar-celeryd-default_1       result = task.revert(**arguments)
sodar-celeryd-default_1     File "/usr/src/app/taskflowbackend/tasks/irods_tasks.py", line 604, in revert
sodar-celeryd-default_1       self.irods.collections.remove(coll_path, recurse=True)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/manager/collection_manager.py", line 70, in remove
sodar-celeryd-default_1       response = conn.recv()
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/irods/connection.py", line 120, in recv
sodar-celeryd-default_1       raise NetworkException("Could not receive server response")
sodar-celeryd-default_1   irods.exception.NetworkException: Could not receive server response
sodar-celeryd-default_1   2022-11-17 09:15:21,909 [ERROR] taskflowbackend.flows: Exception in run_flow(): Could not receive server response
sodar-celeryd-default_1   [2022-11-17 09:15:21,909: ERROR/ForkPoolWorker-15] Exception in run_flow(): Could not receive server response
sodar-celeryd-default_1   [2022-11-17 09:15:21,913: INFO/ForkPoolWorker-15] Unlock OK for project f2acceb7-067d-41a4-8e39-236c022678f1
sodar-celeryd-default_1   2022-11-17 09:15:21,916 [ERROR] taskflowbackend.api: Error running flow: Could not receive server response
sodar-celeryd-default_1   [2022-11-17 09:15:21,916: ERROR/ForkPoolWorker-15] Error running flow: Could not receive server response
sodar-celeryd-default_1   [2022-11-17 09:15:21,939: ERROR/ForkPoolWorker-15] Task taskflowbackend.tasks_celery.submit_flow_task[8a108034-4bc3-4ffb-b3c5-705510bc9ec6] raised unexpected: FlowSubmitException('Error running flow: Could not receive server response')
sodar-celeryd-default_1   Traceback (most recent call last):
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task
sodar-celeryd-default_1       R = retval = fun(*args, **kwargs)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/sentry_sdk/integrations/celery.py", line 204, in _inner
sodar-celeryd-default_1       reraise(*exc_info)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/sentry_sdk/_compat.py", line 56, in reraise
sodar-celeryd-default_1       raise value
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/sentry_sdk/integrations/celery.py", line 199, in _inner
sodar-celeryd-default_1       return f(*args, **kwargs)
sodar-celeryd-default_1     File "/usr/local/lib/python3.8/site-packages/celery/app/trace.py", line 734, in __protected_call__
sodar-celeryd-default_1       return self.run(*args, **kwargs)
sodar-celeryd-default_1     File "/usr/src/app/taskflowbackend/tasks_celery.py", line 29, in submit_flow_task
sodar-celeryd-default_1       taskflow.run_flow(
sodar-celeryd-default_1     File "/usr/src/app/taskflowbackend/api.py", line 186, in run_flow
sodar-celeryd-default_1       cls._raise_flow_exception(ex_msg, tl_event, None)
sodar-celeryd-default_1     File "/usr/src/app/taskflowbackend/api.py", line 59, in _raise_flow_exception
sodar-celeryd-default_1       raise cls.FlowSubmitException(ex_msg)
sodar-celeryd-default_1   taskflowbackend.api.TaskflowAPI.FlowSubmitException: Error running flow: Could not receive server response
```

**Comment by mikkonie** [comment]
The interesting part obviously is this:

```
ValueError: non-zero flags not allowed in calls to recv() on <class 'ssl.SSLSocket'>
```

Why does that happen?

**Comment by mikkonie** [comment]
This **may** have been caused by bihealth/sodar-docker-compose#32 which has now been fixed by #909, but I'm not certain. Monitoring to be sure.

**Comment by mikkonie** [comment]
This has not appeared again in over a month, so I believe it was indeed caused by the aforementioned issue and fixed. Closing.
